libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+  g_autoptr(GVariant) result = g_dbus_proxy_call_sync (proxy,
+                                                       method_name,
+                                                       NULL,
+                                                       G_DBUS_CALL_FLAGS_NONE,
+                                                       -1,
+                                                       cancellable,
+                                                       error);
+  g_autoptr(GVariant) shards_variant = NULL;
+  g_autoptr(GVariant) models_variant = NULL;
+  g_auto(GStrv) shards_strv = NULL;
+  g_autoptr(GPtrArray) model_hts = NULL;
+  GVariantIter iter;
+  GVariant *model_variant = NULL;
+
+  if (result == NULL)
+    return NULL;
+
+  g_variant_get (result, "(^as@aa{ss})", &shards_strv, &models_variant);
+
+  model_hts = g_ptr_array_new_full (g_variant_n_children (models_variant),
+                                    (GDestroyNotify) g_hash_table_unref);
+
+  g_variant_iter_init (&iter, models_variant);
+  while (g_variant_iter_loop (&iter, "@a{ss}", &model_variant))
+    g_ptr_array_add (model_hts,
+                     hashtable_from_variant (model_variant));
+
+  /* Now that we have the models and shards, we can marshal them into
+   * a GPtrArray containing the discovery-feed models */
+  model_data_array = marshal_func ((const gchar * const *) shards_strv,
+                                   model_hts,
+                                   marshal_data);
+
+  return g_steal_pointer (&model_data_array);
+}
+
+static GObject *
+call_dbus_proxy_and_construct_from_model (GDBusProxy           *proxy,
+                                          const gchar          *method_name,
+                                          const gchar          *source_name,
+                                          ModelFromResultFunc   marshal_func,
+                                          gpointer              marshal_data,
+                                          GCancellable         *cancellable,
+                                          GError              **error)
+{
+  g_autoptr(GPtrArray) model_data_array = NULL;
+  g_autoptr(GVariant) result = g_dbus_proxy_call_sync (proxy,
+                                                       method_name,
+                                                       NULL,
+                                                       G_DBUS_CALL_FLAGS_NONE,
+                                                       -1,
+                                                       cancellable,
+                                                       error);
+  g_autoptr(GVariant) model_variant = NULL;
+  g_autoptr(GHashTable) model_ht = NULL;
+
+  if (result == NULL)
+    return NULL;
+
+  g_variant_get (result, "(@a{ss})", &model_variant);
+  model_ht = hashtable_from_variant (model_variant);
+
+  /* Now that we have the model, we can marshal it into a
+   * GObject */
+  return marshal_func (model_ht, marshal_data);
```

Comment:
Need to marshal into a GObject since `EosDiscoveryFeedWordCardStore` does not inherit `EosDiscoveryFeedBaseCardStore` (it doesn't have a card type as it is a child store for EosDiscoveryFeedWordQuoteCardStore)